server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java

```diff
@@ -535,50 +613,248 @@ protected Releasable drainSync() {
+        long minGenerationToBeDeleted = metadataFileGenerationMap.containsKey(oldestMetadataFileToBeDeleted)
+            ? metadataFileGenerationMap.get(oldestMetadataFileToBeDeleted).v1()
+            : getMinMaxTranslogGenerationFromMetadataFile(oldestMetadataFileToBeDeleted, translogTransferManager).v1();
+
+        TreeSet<Tuple<Long, Long>> pinnedGenerations = getOrderedPinnedMetadataGenerations(metadataFileGenerationMap);
+        for (long generation = maxGenerationToBeDeleted; generation >= minGenerationToBeDeleted; generation--) {
+            // Check if the generation is not referred by metadata file matching pinned timestamps
+            if (isGenerationPinned(generation, pinnedGenerations) == false) {
+                generationsToBeDeleted.add(generation);
+            }
         }
+
+        return generationsToBeDeleted;
     }
 
-    /**
-     * Deletes remote translog and metadata files asynchronously corresponding to the generations.
-     * @param generations generations to be deleted.
-     */
-    private void deleteRemoteGeneration(Set<Long> generations) {
-        translogTransferManager.deleteGenerationAsync(
-            primaryTermSupplier.getAsLong(),
-            generations,
-            remoteGenerationDeletionPermits::release
+    // Visible for testing
+    protected List<String> getMetadataFilesToBeDeleted(List<String> metadataFiles) {
+        Tuple<Long, Set<Long>> pinnedTimestampsState = RemoteStorePinnedTimestampService.getPinnedTimestamps();
+
+        // Keep files since last successful run of scheduler
+        List<String> metadataFilesToBeDeleted = RemoteStoreUtils.filterOutMetadataFilesBasedOnAge(
+            metadataFiles,
+            file -> RemoteStoreUtils.invertLong(file.split(METADATA_SEPARATOR)[3]),
+            pinnedTimestampsState.v1()
+        );
+
+        // Get md files matching pinned timestamps
+        Set<String> implicitLockedFiles = RemoteStoreUtils.getPinnedTimestampLockedFiles(
+            metadataFilesToBeDeleted,
+            pinnedTimestampsState.v2(),
+            metadataFilePinnedTimestampMap,
+            file -> RemoteStoreUtils.invertLong(file.split(METADATA_SEPARATOR)[3]),
+            TranslogTransferMetadata::getNodeIdByPrimaryTermAndGen
         );
+
+        // Filter out metadata files matching pinned timestamps
+        metadataFilesToBeDeleted.removeAll(implicitLockedFiles);
+
+        return metadataFilesToBeDeleted;
+    }
+
+    // Visible for testing
+    protected boolean isGenerationPinned(long generation, TreeSet<Tuple<Long, Long>> pinnedGenerations) {
+        Tuple<Long, Long> ceilingGenerationRange = pinnedGenerations.ceiling(new Tuple<>(generation, generation));
+        if (ceilingGenerationRange != null && generation >= ceilingGenerationRange.v1() && generation <= ceilingGenerationRange.v2()) {
+            return true;
+        }
+        Tuple<Long, Long> floorGenerationRange = pinnedGenerations.floor(new Tuple<>(generation, generation));
```

Comment:
why do we need ceil and floor checks both ?

Reply:
Assume we have following generation ranges in metadata files:

```
min=3, max=42, 
min=23, max=82, 
min=63, max=122, 
min=124, max=163, 
min=144, max=203, 
min=184, max=243, 
min=224, max=283, 
min=264, max=323, 
min=304, max=363, 
min=344, max=403
```

Let's take an example of 2 generations: `140` and `160`.
For both of these generations, floor is `<124, 163>` and ceiling is `<144, 203>`. 

For `140`, we need floor check in order to understand the generation is pinned.
For `160`, we need ceiling check in order to understand the generation is pinned.

Reply:
from what i understand the goal here to find out of the `generation` falls b/w any one of the `pinnedGenerations` ranges?

Reply:
> from what i understand the goal here to find out of the generation falls b/w any one of the pinnedGenerations ranges?

Yes